### DISCUSSION TITLE
generating audit log automatically in Go

### DISCUSSION THREAD

**Discussion by jkdoshi** [pocketbase/pocketbase#1281] [Q&A] [answered]
I am using PB as Go framework and am trying to auto-populate a table of records tracking which table had which record insert/update/delete'd when and by which user.

* created a table named `auditlog` with fields `table`, `event` (insert/update/delete), `user` (the system field `created` comes for free)
* In `OnBeforeServe` hook register a listener for `OnModelAfterCreate` (QUESTION: is that the best approach? What's the difference between `OnModel*` and `OnRecord*` hooks?)
* The listener is invoked whenever record is insert/update/delete'd. So insert a record into the `auditlog` table each time.

PROBLEM: In the listener above, I have all the record information except identity of the user for the request. How do I get that?

Also, any suggestions for a good design for this would be useful. I think there is this or something like this in the roadmap.

**Comment by ganigeorgiev** [comment, marked as answer]
Is `autditlog` some plain table or collection? If it is a collection it would be easier. Here is an example (_I haven't tested it_):

```go
package main

import (
	"log"

	"github.com/labstack/echo/v5"
	"github.com/pocketbase/pocketbase"
	"github.com/pocketbase/pocketbase/apis"
	"github.com/pocketbase/pocketbase/core"
	"github.com/pocketbase/pocketbase/models"
)

func insertAuditlog(
	app core.App,
	httpContext echo.Context,
	record *models.Record,
	event string,
) error {
	if record.Collection().Name == "auditlog" {
		return nil // ignore auditlog record changes
	}

	auditlogCollection, err := app.Dao().FindCollectionByNameOrId("auditlog")
	if err != nil {
		return err
	}

	var adminId string
	var authRecordId string

	// get the authenticated admin
	admin, _ := httpContext.Get(apis.ContextAdminKey).(*models.Admin)
	if admin != nil {
		adminId = admin.Id
	}

	// or get the authenticated user record
	authRecord, _ := httpContext.Get(apis.ContextAuthRecordKey).(*models.Record)
	if authRecord != nil {
		authRecordId = authRecord.Id
	}

	auditlog := models.NewRecord(auditlogCollection)
	auditlog.Set("table", record.Collection().Name)
	auditlog.Set("record", record.Id)
	auditlog.Set("event", event)
	auditlog.Set("user", authRecordId)
	auditlog.Set("admin", adminId)

	return app.Dao().SaveRecord(auditlog)
}

func main() {
	app := pocketbase.New()

	app.OnRecordAfterCreateRequest().Add(func(e *core.RecordCreateEvent) error {
		return insertAuditlog(app, e.HttpContext, e.Record, "create")
	})

	app.OnRecordAfterUpdateRequest().Add(func(e *core.RecordUpdateEvent) error {
		return insertAuditlog(app, e.HttpContext, e.Record, "update")
	})

	app.OnRecordAfterDeleteRequest().Add(func(e *core.RecordDeleteEvent) error {
		return insertAuditlog(app, e.HttpContext, e.Record, "delete")
	})

	if err := app.Start(); err != nil {
		log.Fatal(err)
	}
}
```

---

> is that the best approach?

Yes, it should be fine. Using the hook is the recommended way to extend PocketBase.

---

> What's the difference between OnModel* and OnRecord* hooks?

`OnModel*` hooks fire on every model change (not just `models.Record`) when it is inserted/updated/deleted with the `Dao()` service (even on cascade update/delete) (aka. on app initiated db change -> you'll have a model event).

`OnRecord*` hooks (and every other event hook that ends with `*Request`) fires **only** on HTTP API requests. `OnRecord*` hooks also could trigger one or many `OnModel*` hooks.
Also please note that only the `*Request` ending hooks have access to the client request context (aka. `e.HttpContext`).

It is not very clear from the site docs, but you can find more info when a hook is triggered by expanding the related hook "accordion" panel - https://pocketbase.io/docs/event-hooks/.

---

> In the listener above, I have all the record information except identity of the user for the request. How do I get that?

You can get it from the `echo.Context` request instance. Depending from where (route, middleware, hook), there are several ways you can access it. You can find examples in https://pocketbase.io/docs/router/#get-logged-admin-or-app-user

**Reply by hattorihanzo86** [reply]
Im currently implementing this what is the filename / directory to modified or added this?

**Reply by ganigeorgiev** [reply]
@hattorihanzo86 The above code is a standalone Go program. You can name the file or restructure it however you like.
There will be more detailed documentation about this with v0.17.0 (_including option to bind to the app hooks with JS_), but for now you can check https://pocketbase.io/docs/use-as-framework/.

**Reply by jkdoshi** [reply]
@hattorihanzo86 you can see a fully implemented example of automatically generated auditlog in the [pocketbase-sveltekit-starter](https://github.com/spinspire/pocketbase-sveltekit-starter) project [here](https://github.com/spinspire/pocketbase-sveltekit-starter/blob/master/pb/auditlog/auditlog.go).

**Reply by hattorihanzo86** [reply]
So its not possible on the repo that im working in the master branch to implement that?

**Reply by hattorihanzo86** [reply]
Thanks but how can i get the newValue example oldValue =2, newValue =4
i have newValue in collection i wanted to show what they change showing oldValue and newValue
![image](https://github.com/pocketbase/pocketbase/assets/4376706/3c3dffcc-3d06-411c-94a5-68625f99df35)

```

func insertAuditlog(
	app core.App,
	httpContext echo.Context,
	record *models.Record,
	event string,
) error {
	if record.Collection().Name == "auditlog" {
		return nil // ignore auditlog record changes
	}

	auditlogCollection, err := app.Dao().FindCollectionByNameOrId("auditlog")
	if err != nil {
		return err
	}

	var adminId string
	var authRecordId string

	// get the authenticated admin
	admin, _ := httpContext.Get(apis.ContextAdminKey).(*models.Admin)
	if admin != nil {
		adminId = admin.Id
	}

	// or get the authenticated user record
	authRecord, _ := httpContext.Get(apis.ContextAuthRecordKey).(*models.Record)
	if authRecord != nil {
		authRecordId = authRecord.Id
	}

	auditlog := models.NewRecord(auditlogCollection)
	auditlog.Set("table", record.Collection().Name)
	auditlog.Set("record", record.Id)
	auditlog.Set("event", event)
	auditlog.Set("user", authRecordId)
	auditlog.Set("admin", adminId)
	
	return app.Dao().SaveRecord(auditlog)
}

```

**Reply by ganigeorgiev** [reply]
@hattorihanzo86 You can define 2 `json` columns in the `auditlog` collection (eg. `old` and `new`) and store respectively the old and new record schema fields data. The new record schema fields data could be exported with [`record.SchemaData()`](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.16.10/models#Record.SchemaData), The old record schema data could be exported using [`record.OritinalCopy().SchemaData()`](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.16.10/models#Record.OriginalCopy). If you want a diff you'll have to compare the 2 data maps on your own.

**Reply by hattorihanzo86** [reply]
thank you will resume when i got back in my desk thanks so much

**Comment by jkdoshi** [comment]
As usual, very impressed with your prompt and detailed responses. Your solution worked 100% for me.

One question though ... in `event.go` I see the following:

```go
type RecordCreateEvent struct {
	HttpContext echo.Context
	Record      *models.Record
}

type RecordUpdateEvent struct {
	HttpContext echo.Context
	Record      *models.Record
}

type RecordDeleteEvent struct {
	HttpContext echo.Context
	Record      *models.Record
}
```
I don't doubt that even though the above three look exactly the same, you still need 3 different types. But why not just have 3 types derived from the same base type? As a consequence of the above, I had to create 3 separate functions that differ only in their argument type to handle insert, update, and delete events.

Given how well you've designed the platform so far, I feel there must be a good reason for it that I just can't see.

Once again, thank you for working on PocketBase.

**Reply by ganigeorgiev** [reply]
The idea was that the event structs may change eventually by defining request action specific data (there were plans also to add the "old/original" record data to the create and update events, but later I decided to make it part of `models.Record.OriginalCopy()` so that is no longer necessary).

In the future I may consider merging them but that will be at later stage.

**Comment by jkdoshi** [comment]
So here's a followup on my auditlog saga.

Here's what my auditlog collection looks like:
![image](https://github.com/pocketbase/pocketbase/assets/2412869/aface4e8-e0d5-452a-8b65-23c757a5f1a7)

I'm able to generate auditlog table as described above. But I'm not able to secure it just right. I have two choices:

- Make access admin-only
- Give everyone (or all authenticated users) List and View access to all records

First one is too closed. Second one is too open.

Is there a way I could add custom View/List rule in "API Rules" or even in the Go backend to:
- ensure that a user has access to an auditlog record ONLY if they have access to the corresponding original record in {{collection}} collection with {{record}} ID

So, for example, if in the "auditlog" there a record `{"collection": "customers", "record": "xyz", ...}`, then give view/list access only if the user has view/list access to "xyz" record of "customers" collection.

How would I implement it using API rules in admin UI or `app.OnRecordsListRequest` Go API?

**Reply by ganigeorgiev** [reply]
What you want can't be implemented with just the API rules from the Admin UI (_or with a custom API rule, at least I can't think of a performant way how this can be done_).

The best approach probably would be to register a custom route and use the new [`Dao.CanAccessRecord()` helper](https://github.com/pocketbase/pocketbase/blob/d9e1a759a1f1ff58e611e8af79657ff85d26df63/daos/record.go#L512-L566) (_will be available with the upcoming v0.17.0 release_) **for each record**.

Another possible workaround since you want to give auditlog access to the same user that can view/list the original record, it may be easier to instead of having a separate collection, to create a "versions/history" `json` field in each original collection and maintain it with Go with every update record request (You can export the record fields data by calling `record.SchemaData()`).

_Edit: Note that for the above both cases the workarounds will not work if the record was deleted._
